cedar-policy-validator/src/entity_manifest.rs

```diff
@@ -0,0 +1,1103 @@
+            *root_trie = root_trie.union(&flat_slice.to_primary_slice());
+        }
+        ExprKind::GetAttr { .. } => {
+            let flat_slice = get_expr_path(expr, policy_id)?;
+
+            // PANIC SAFETY: Successfuly typechecked expressions should always have annotated types.
+            #[allow(clippy::expect_used)]
+            let leaf_field = if should_load_all {
+                type_to_access_trie(
+                    expr.data()
+                        .as_ref()
+                        .expect("Typechecked expression missing type"),
+                )
+            } else {
+                AccessTrie::new()
+            };
+
+            *root_trie = flat_slice.to_primary_slice_with_leaf(leaf_field);
+        }
+    };
+
+    Ok(())
+}
+
+/// Compute the full [`AccessTrie`] required for the type.
+fn type_to_access_trie(ty: &Type) -> AccessTrie {
+    match ty {
+        // if it's not an entity or record, slice ends here
+        Type::ExtensionType { .. }
+        | Type::Never
+        | Type::True
+        | Type::False
+        | Type::Primitive { .. }
+        | Type::Set { .. } => AccessTrie::new(),
+        Type::EntityOrRecord(record_type) => entity_or_record_to_access_trie(record_type),
+    }
+}
+
+/// Compute the full [`AccessTrie`] for the given entity or record type.
+fn entity_or_record_to_access_trie(ty: &EntityRecordKind) -> AccessTrie {
+    match ty {
+        EntityRecordKind::ActionEntity { name: _, attrs }
+        | EntityRecordKind::Record {
+            attrs,
+            open_attributes: _,
+        } => {
+            let mut fields = HashMap::new();
+            for (attr_name, attr_type) in attrs.iter() {
+                fields.insert(
+                    attr_name.clone(),
+                    Box::new(type_to_access_trie(&attr_type.attr_type)),
+                );
+            }
+            AccessTrie {
+                children: fields,
+                parents_required: false,
+                data: (),
+            }
+        }
+
+        EntityRecordKind::Entity(_) | EntityRecordKind::AnyEntity => {
+            // no need to load data for entities, which are compared
+            // using ids
+            AccessTrie::new()
+        }
+    }
+}
+
+/// Given an expression, get the corresponding data path
+/// starting with a variable.
+/// If the expression is not a `<datapath-expr>`, return an error.
+/// See [`FailedAnalysisError`] for more information.
+fn get_expr_path(
+    expr: &Expr<Option<Type>>,
+    policy_id: &PolicyID,
+) -> Result<AccessPath, EntityManifestError> {
+    Ok(match expr.expr_kind() {
+        ExprKind::Slot(slot_id) => {
+            if slot_id.is_principal() {
+                AccessPath {
+                    root: EntityRoot::Var(Var::Principal),
+                    path: vec![],
+                    parents_required: false,
+                }
+            } else {
+                assert!(slot_id.is_resource());
```

Comment:
instead of this if-else with an `assert!`, should we just `match` on the `slot_id`?

Reply:
I think I can't access the inner enum from this crate